File: javascript/clock/clock.js
const HOUR_MINUTES = 60;
const DAY_MINUTES = 24 * 60;

const modulus = (time_as_minutes) => (DAY_MINUTES + (time_as_minutes % DAY_MINUTES)) % DAY_MINUTES;

export class Clock {
  constructor(hours, minutes = 0) {
    this.time_as_minutes = modulus(hours * HOUR_MINUTES + minutes);
  }

  calculateMinutes() {
    return this.time_as_minutes % HOUR_MINUTES;
  }

  calculateHours() {
    return Math.floor(this.time_as_minutes / HOUR_MINUTES);
  }

  toString() {
    return [this.calculateHours(), this.calculateMinutes()].map(time => `${time}`.padStart(2, '0')).join(':');
  }

  plus(minutes) {
    return new Clock(0, this.time_as_minutes + minutes);
  }

  minus(minutes) {
    return this.plus(-minutes);
  }

  equals(clock) {
    return this.time_as_minutes == clock.time_as_minutes;
  }
}

File: ruby/clock/clock.rb
class Clock
  HOUR_MINUTES = 60
  DAY_MINUTES = 1440

  ABSOLUTE = ->(hour, minute){((hour * HOUR_MINUTES) + minute) % DAY_MINUTES}

  def initialize(hour: 0, minute: 0)
    @absolute_minutes = ABSOLUTE.call(hour, minute)
  end

  attr_reader :absolute_minutes

  def to_s
    format("%02d:%02d", calculate_hours, calculate_minutes)
  end

  def +(clock)
    Clock.new(minute: absolute_minutes + clock.absolute_minutes)
  end

  def -(clock)
    Clock.new(minute: absolute_minutes - clock.absolute_minutes)
  end

  def ==(clock)
    absolute_minutes == clock.absolute_minutes
  end

  private

  def calculate_hours
    (absolute_minutes / HOUR_MINUTES).round
  end

  def calculate_minutes
    absolute_minutes % HOUR_MINUTES
  end
end
